### DISCUSSION TITLE
Help with usage

### DISCUSSION THREAD

**Discussion by RedHappyLlama** [JaffaKetchup/flutter_map_tile_caching#51] [Q&A] [answered]
Hi,

Within the readme it's stated "Please note that this library has not been tested on iOS devices, so issues may arise. Please leave an issue if they do, and I'll try my best to debug and solve them.".

Just curious if this still the case? 

Thanks!

**Comment by JaffaKetchup** [comment, marked as answer]
Yeah, unfortunately I don't have a Mac to test things on, and running a VM is out of the question for me right now.

Thanks for your interest! Still happy to try to help, but I may be limited...

**Reply by RedHappyLlama** [reply]
Thanks,

I have been able to run the example app on my mac and iPhone, with the changes suggested here: #49. However I'm struggling to understand how to use your amazing package (I have only been coding for a year or 2). 

I think I've understood how to download a set of tiles:

`    tileProvider ??= StorageCachingTileProvider.fromMapCachingManager(mcm!);
    mapDownload ??= tileProvider!.downloadRegion(
      CircleRegion(LatLng(53.6, -4.0), 0.5).toDownloadable(
        12,
        13,
        TileLayerOptions(
          urlTemplate: "http://b.tile.openstreetmap.fr/hot/{z}/{x}/{y}.png",
          subdomains: ['a', 'b', 'c'],
        ),
        parallelThreads: 2,
        preventRedownload: false,
        seaTileRemoval: false,
      ),
    ).asBroadcastStream();`

However I'm really struggling. Do you have a much more simplified example app you can share to help? All I'm trying to do is create the following user story:

1. There is a list of various different maps (with a very small number of tiles) that are available for download, with a LatLng centre and radius defined.
2. The user selects which map to download.
3. The download occurs.
4. User can then open the downloaded map, without any internet connection. In an ideal world, if the user happens to have an internet connection when the map is being used, other tiles outside of the downloaded region are shown.
5. The user can select other maps to download when they have an internet connection from the list.
6. Ability to delete one or all downloaded maps.

I think a simplified example app will really help people understand the basics of your package, and then explore the advanced features on their own. This will then make your package even more popular and successful! 

Best Regards!

**Reply by JaffaKetchup** [reply]
Thanks for your feedback, I appreciate it!

Complexity has been an issue throughout this project, due its wide functionality coverage. This is particularly due to the unclear syntax (which I'm working on fixing for v5); however it's good to see you've figured it out :).
In terms of a simpler example app, I unfortunately don't have one to share. At the moment, writing a new one is not really an option, as v5 syntax is completely different to the current v4; but I will likely simplify the example when I've finalised the v5 syntax.

However, your request seems relatively easy to implement:

- As you've got regions that you want the user to be able to choose between, I'd recommend setting up a list of `CircleRegions` at your app load time. This is inexpensive, so won't impact performance.
- Not sure how your app is laid out, but have an app bar button or settings menu, similar to the example app. Inside here you can have a menu, where each item represents a cache store with the name of the respective circle region's coordinates. When tapped, use a state management solution to change a currently active variable to the new store. If the map has not been downloaded, use your script above here: remember to attach a listener or it might not do anything. If the user holds down on the item, empty and delete the store.
- When using the `StorageCachingTileProvider`, ensure the cache behaviour option (can't remember exactly what it is right now, sorry) is NOT set to cache only. This should achieve the effect you want. Note that any tiles they do browse over will be cached: there's no way around this at the moment, but I'll try adding it if you need it.

I hope you can follow along! I won't be able to help now until about 4pm tomorrow (BST), but I'm happy to help again after that!

Just out of interest, if you have some spare time: do you have any ideas for an easier syntax for v5? Currently I'm migrating it to something like `FlutterMapTileCaching.instance['store name'].download.start()` and `FMTC.instance.stats.allStoreSizes`, where `FlutterMapTileCaching` and `FMTC` are equivalent and initialised at startup with a really easy call (to take away manual state management of things like the root directory). What are your thoughts on this? I'd be grateful for your input!

**Reply by RedHappyLlama** [reply]
Hi,

V5 sounds good, any ideas for a release date? I also like the proposed new syntax, looks like it will make it cleaner and simpler although difficult to judge based on little time with the package and out of context. 

In terms of your second point, is this a way of creating a new cache store and downloading the tiles to that new store (please excuse quality of the code, trying to keep it simple and within my skills). 

Have I also attached a listener and used it correctly (obviously give you the ability to react and update the state accordingly)? 

```
void _downloadMap({required int? mapID}) {
    MapCachingManager mcm = MapCachingManager(parentDirectory!, '${mapID}MapDownload');
    tileProvider ??= StorageCachingTileProvider.fromMapCachingManager(mcm);
    mapDownload ??= tileProvider!.downloadRegion(
      CircleRegion(LatLng(53.6, -4.0), 0.5).toDownloadable(
        12,
        13,
        TileLayerOptions(
          urlTemplate: "http://b.tile.openstreetmap.fr/hot/{z}/{x}/{y}.png",
          subdomains: ['a', 'b', 'c'],
        ),
        parallelThreads: 2,
        preventRedownload: false,
        seaTileRemoval: false,
      ),
    ).asBroadcastStream();

    mapDownload!.listen((event) {
      print('map download progress is: ${event.percentageProgress}');
    }, onError: (event) {
      print('map download encountered an error');
    }, onDone: () {
      print('map download complete');
    }, cancelOnError: true);
  }
```

and then to use that store in a FlutterMap:

```
              TileLayerOptions(
                urlTemplate: "http://b.tile.openstreetmap.fr/hot/{z}/{x}/{y}.png",
                subdomains: ['a', 'b', 'c'],
                tileProvider: StorageCachingTileProvider(
                  behavior: CacheBehavior.cacheFirst,
                  parentDirectory: widget.parentDirectory,
                  storeName: '${widget.mapId}MapDownload'
                )
              ),
```

Your help is really appreciated! Sorry for being such a noob, still learning :).

**Reply by JaffaKetchup** [reply]
Hi there,

Unfortunately, there is no planned release date. Hopefully you can migrate to v5 easily once you implement v4. Your feedback is encouraging, thanks :)

There is a way to make a new store and download to it, just constructing the appropriate `MapCachingManager` or `StorageCachingTileProvider` should do it automatically. Then you can use it as normal. If this isn't working, maybe something else is wrong?

In terms of attaching a listener, you seem to have done it correctly. Note that there should never be an error through the `onError` callback you've defined. Errors are always described in the `DownloadProgress` object.

You also appear to have setup the tile provider correctly.

Don't worry about needing help, everyone has to start somewhere, and I haven't made this the easiest library to start out learning Flutter with!

**Reply by RedHappyLlama** [reply]
Hi, 

Thanks for your help, I've got some working code I can start to build. Really appreciate it! 

One final question, without the changes shown #49, the package doesn't work on iOS. I've managed to make the changes myself to a local version of the package so the example project works. However in the app I'm developing I've no idea how to do this so the app works. i.e. I don't know how to use a package that isn't released. Are you able to help?

If I can repay you in anyway someone to help out as I've got a mac and iPhone, I'd be happy to. 

Thanks!

**Reply by JaffaKetchup** [reply]
That's great! To use someone else's fork, you can use a 'dependency_overrides' section in your pubspec.yaml. Try this:

```yaml
dependencies:
    flutter_map_tile_caching: ^4.0.1

dependency_overrides:
    flutter_map_tile_caching:
        git:
            url: https://github.com/ristiisa/flutter_map_tile_caching.git
            ref: watcher
```

I think that should work...

Thanks for the offer, but don't worry about it :)

Happy to help more!

PS. If this package is helping you build a commercial app, please consider donating if you can, it really helps. If you can't, a star and a like is also greatly appreciated!

**Reply by JaffaKetchup** [reply]
Hi @RedHappyLlama, just following up, did you manage to get everything you needed done?

**Reply by RedHappyLlama** [reply]
Liked and stared! If I get to a commercial level I'll be sure the donate. 

Early days with the code, but I think I have everything I need for now. Many thanks again for your help! Super star!

**Reply by JaffaKetchup** [reply]
@RedHappyLlama Just updating you, v5 has now been released. Happy to help if you need it!

**Reply by RedHappyLlama** [reply]
@JaffaKetchup thanks very much for the update, as and when I update I'll be in touch if I need any help!

**Reply by RedHappyLlama** [reply]
Hi @JaffaKetchup,

Just tried migrating to v5 and I think I've got a slight issue. I can download tiles fine in v5. However, it doesn't seem to be downloading different layers of tiles for the same region now I'm on v5. It seems to download just one layer.

For both the MapOptions max zoom and min zoom were 12 and 16 which means you weren’t able to zoom out of the downloaded map. However now on v5 it is currently possible to zoom out of the downloaded map, whilst staying inside max and min zoom constraints.

Previous code for v4:
Download:
```
  Future _downloadMap() async {
    MapCachingManager mcm = MapCachingManager(parentDirectory!, '${id}MapDownload');

    tileProvider = StorageCachingTileProvider.fromMapCachingManager(mcm);
    mapDownload = tileProvider!
        .downloadRegion(
          RectangleRegion(bounds).toDownloadable(
            12,
            16,
            TileLayerOptions(
              urlTemplate: "http://b.tile.openstreetmap.fr/hot/{z}/{x}/{y}.png",
              subdomains: ['a', 'b', 'c'],
            ),
            parallelThreads: 2,
            preventRedownload: false,
            seaTileRemoval: false,
          ),
        )
        .asBroadcastStream();

    mapDownload!.listen((event) {
      print('map download progress is: ${event.percentageProgress}');
    }, onError: (event) {
      print('map download encountered an error');
    }, onDone: () {
      print('map download complete');
    }, cancelOnError: true);

    Future downloadComplete = mapDownload!.firstWhere((element) => element.percentageProgress == 100);
    return downloadComplete;
  }
```


Use:
```
            layers: [
              TileLayerOptions(
                  urlTemplate: "http://b.tile.openstreetmap.fr/hot/{z}/{x}/{y}.png",
                  subdomains: ['a', 'b', 'c'],
                  tileProvider: StorageCachingTileProvider(
                      behavior: CacheBehavior.cacheFirst,
                      parentDirectory: context.read<DownloadController>().parentDirectory!,
                      storeName: '${widget.id}MapDownload')),


```
Current code for v5:
Download:
```
  Future _downloadMap({required LatLngBounds bounds}) async {
    print('downloading map');
    StoreDirectory store = FMTC.instance('${id}MapDownload');
    await store.manage.createAsync();
    mapDownload = store.download
        .startForeground(
          region: RectangleRegion(bounds).toDownloadable(
            12,
            16,
            TileLayerOptions(
              urlTemplate: "http://b.tile.openstreetmap.fr/hot/{z}/{x}/{y}.png",
              subdomains: ['a', 'b', 'c'],
            ),
            parallelThreads: 2,
            preventRedownload: false,
            seaTileRemoval: false,
          ),
        )
        .asBroadcastStream();
    mapDownload!.listen((event) {
      print('map download progress is: ${event.percentageProgress}');
    }, onError: (event) {
      print('map download encountered an error');
    }, onDone: () {
      print('map download complete');
    }, cancelOnError: true);
    Future downloadComplete = mapDownload!.firstWhere((element) => element.percentageProgress == 100);
    return downloadComplete;
  }
```
Use:
```
            layers: [
              TileLayerOptions(
                  urlTemplate: "https://b.tile.openstreetmap.fr/hot/{z}/{x}/{y}.png",
                  subdomains: ['a', 'b', 'c'],
                  tileProvider: FMTC.instance('${widget.id}MapDownload').getTileProvider()),

```

**Reply by JaffaKetchup** [reply]
Hi again @RedHappyLlama,
I can't seem to reproduce this issue. Just to clarify, you mean you can't seem to download multiple zoom levels within the same region at once? If so, can you try the example app? It seems to work for me :)

**Reply by RedHappyLlama** [reply]
Hi @JaffaKetchup,

That's correct, it is either not downloading multiple zoom levels (12 - 16), or can't access them when the phone is fully offline and can only access one download level. 

When you tried to reproduce the issue, did you use my code or the example app? Thanks!

**Reply by JaffaKetchup** [reply]
Ok interesting. I'm using the example app, just because I can't test any code right now.

You may be able to self diagnose if you are running on a desktop platform. On Windows, find and open the 'fmtc' directory located within the Documents directory. Open the store within it, then the 'tiles' directory, and see if you can find any tiles with the zoom levels you are missing.

**Reply by RedHappyLlama** [reply]
Hi, 

I don't have a windows laptop sadly. You can't see anything obviously wrong with my code? I'll try the example app when I can.

**Reply by RedHappyLlama** [reply]
Hi,

I've tried the example app and agree it's downloading multiple zoom levels. However I've been unable to use that to understand why mine is not, but I haven't found it easy to follow the example app so possibly missing something. Any help you can give would be much appreciated. The bit I don't really understand is, I've not changed much to migrate. Thanks!

**Reply by RedHappyLlama** [reply]
Hi, 

I've cracked it. I realised the zoom level I thought it was "downloading" was actually just the visible tiles at the time being cached. The tiles actually being downloaded weren't being shown. I realised the root cause was that for the urlTemplate to download I was only using http, but the urlTemplate to view I was using http***s***. Seemingly this difference meant the downloaded tiles weren't displaying.

**Reply by JaffaKetchup** [reply]
Ah, glad to hear you fixed it, and sorry for the confusion!

**Comment by mziadd** [comment]
It's working fine on apple devices ipad/iphone
